onnx/defs/tensor/defs.cc

```diff
@@ -8,6 +8,174 @@
+    block_num_element *= effective_block_shape[i];
+  }
+
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if (pads.size() != n_input_dims * 2) {
+      fail_shape_inference("Attribute pads has incorrect size");
+    }
+  } else {
+    pads.assign(n_input_dims * 2, 0);
+  }
+
+  auto output_shape =
+      ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+  *output_shape->add_dim() = input_shape.dim(0);
+  auto newdim = output_shape->add_dim();
+  if (input_shape.dim(1).has_dim_value()) {
+	  block_num_element *= input_shape.dim(1).dim_value();
+	  newdim->set_dim_value(block_num_element);
+  }
+
+  size_t last_dim_size = 1;
+  newdim = output_shape->add_dim();
+  int effective_block_shape_size = static_cast<int>(effective_block_shape.size());
+  for (int i = 0; i < effective_block_shape_size; ++i) {
+    if (!input_shape.dim(2 + i).has_dim_value()) {
+      return;
+    }
+    // how big is the input, including padding
+    int64_t effective_input_size = input_shape.dim(2 + i).dim_value();
+    effective_input_size += pads[i];
+    effective_input_size += pads[i + effective_block_shape_size];
+
+    // how many times we can move the kernel from it's initial position, based
+    // on the stride
+    int64_t strided_kernel_positions;
+    strided_kernel_positions =
+        (effective_input_size - effective_block_shape[i]) / strides[i];
+
+    // add in the initial position
+    last_dim_size *= (1 + strided_kernel_positions);
+  }
+  newdim->set_dim_value(last_dim_size);
+
+}
+
+std::function<void(OpSchema&)> UnfoldToDepthOpSchemaGenerator() {
+  return [=](OpSchema& schema) {
+    std::string doc = R"DOC(
+The UnfoldToDepth operator extracts sliding blocks from an input tensor, and concatenates these blocks
+in the last dimension.
+Given an input of shape (N x C x D1 x D2 ... x Dn), output would be a 3-D tensor of shape:<br/>
+
+```(N, C * reduce-mul(block_size), reduce-mul(num_blocks))```
+<br/>
+Where number of blocks extracted from each spatial dimension d is:
+```
+block_size[d] = floor((input_spatial_shape[d] + 2 * padding[d] − dilation[d] * (kernel_size[d] − 1) − 1) / stride[d]) + 1
+```
+)DOC";
+    schema.SetDoc(doc);
+    schema.Input(
+        0,
+        "X",
+        "Input data tensor from previous layer; "
+        "has size (N x C x H x W), where N is the batch size, "
+        "C is the number of channels, and H and W are the "
+        "height and width. Note that this is for the 2D image. "
+        "Otherwise the size is (N x C x D1 x D2 ... x Dn). ",
+        "T");
+    schema.Output(
+        0,
+        "Y",
+        "Output data tensor that contains the result of the "
+        "unfold. The output has three dimensions, and dimension "
+        "values are funciton of the kernel size, stride size, and "
+        "pad lengths.",
+        "T");
+    schema.TypeConstraint(
+        "T",
+        {"tensor(float16)", "tensor(float)", "tensor(double)"},
```

Comment:
for image, I think uint8 should be supported, please check whether you want to have other numeric types support.